sklearn/linear_model/coordinate_descent.py

```diff
@@ -729,6 +737,9 @@ def fit(self, X, y, check_input=True):
         self.coef_, self.dual_gap_ = map(np.squeeze, [coef_, dual_gaps_])
         self._set_intercept(X_offset, y_offset, X_scale)
 
+        # workaround since _set_intercept will cast self.coef_ into float64
+        self.coef_ = np.asarray(self.coef_, dtype=X.dtype)
```

Comment:
So:

```python
from sklearn.linear_model.base import _preprocess_data
X = rng.randn(5, 5).astype(np.float32)
y = rng.randn(5).astype(np.float32)
for fit_intercept in [True, False]:
    X, y, X_offset, y_offset, X_scale = _preprocess_data(X, y, fit_intercept=True)
    X_scale.dtype
    dtype('float64')
```

I suppose we can do away with this "workaround" if `X_scale` is set to the right dtype in `_preprocess_data`, i.e replace `X_scale = np.ones(X.shape[1])` with `X_scale = np.ones(X.shape[1], dtype=X.dtype)` etc wherever applicable.